### ISSUE TITLE
Protractor is throwing an error when executing more than two tests with restartBrowserBetweenTests set to true

### ISSUE THREAD

**Issue by mm7163** [angular/protractor#2001]
I am using Protractor 2.0.0 and Chrome 41.0.2272.101.

config.js:
```
exports.config = {
    directConnect: true,
    specs: ['testCases/**.js'],
    allScriptsTimeout: 10000,
    restartBrowserBetweenTests: true,
    capabilities: {'browserName': 'chrome'},
    baseUrl: 'http://angularjs.org',
    jasmineNodeOpts: {showColors: true,isVerbose: true,includeStackTrace: true},onPrepare: function() {require('jasmine-reporters');jasmine.getEnv().addReporter(new jasmine.JUnitXmlReporter('./xmloutput', true, true));
    }
};
```
pageObjects/generalObjects.js:
```
var m2testWebGeneralObjects = function() {
    this.text1 = by.xpath('/html/body/div[1]/div[2]/div/div[2]/p');
};
module.exports = new m2testWebGeneralObjects();
```
testCases/test1.js (same as test2.js and test3.js):
```
var m2generalObjects = require('../pageObjects/generalObjects.js');

describe('Test Error', function() {
    var ptor;
    browser.driver.manage().window().maximize();

    beforeEach(function () {
        ptor = browser;
    });

    it('Should display correct content', function() {
        ptor.manage().deleteAllCookies();
        ptor.get(ptor.baseUrl);
        expect(element(m2generalObjects.text1).getText()).toContain('address');
    });
});
```
When I run protractor config.js, after the test1.js is executed I get following error message:
```
c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\webdriver.js:392
      throw new Error('This driver instance does not have a valid session ID ' +

            ^
Error: This driver instance does not have a valid session ID (did you call WebDriver.quit()?) and may no longer be used.
    at checkHasNotQuit (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\webdriver.js:392:13)
    at flow.execute.then.value (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\webdriver.js:365:5)
    at [object Object].promise.ControlFlow.runInFrame_ (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\promise.js:1877:20)
    at [object Object].promise.ControlFlow.runEventLoop_ (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\promise.js:1755:8)
    at [object Object].<anonymous> (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\promise.js:2056:12)
    at [object Object].goog.async.run.processWorkQueue [as _onTimeout] (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\selenium-webdriver\lib\goog\async\run.js:125:21)
    at Timer.listOnTimeout [as ontimeout] (timers.js:110:15)From: Task: WebDriver.manage().window().maximize()
    at [object Object].webdriver.WebDriver.schedule (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\webdriver.js:360:15)
    at [object Object].webdriver.WebDriver.Window.maximize (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\webdriver.js:1474:23)
    at [object Object].<anonymous> (d:\protractor\protractorIssue\testCases\test3.js:6:38)
    at [object Object].jasmine.Env.describe_ (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\minijasminenode\lib\jasmine-1.3.1.js:913:21)
    at [object Object].jasmine.Env.describe (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\minijasminenode\lib\jasmine-1.3.1.js:898:15)
    at describe (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\minijasminenode\lib\jasmine-1.3.1.js:658:27)
    at Object.<anonymous> (d:\protractor\protractorIssue\testCases\test3.js:4:1)

    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
From: Task: WebDriver.manage().timeouts().setScriptTimeout(10000)
    at [object Object].webdriver.WebDriver.schedule (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\webdriver.js:360:15)
    at [object Object].webdriver.WebDriver.Timeouts.setScriptTimeout (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\webdriver.js:1367:23)
    at [object Object].Runner.createBrowser (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\lib\runner.js:181:30)
    at c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\lib\runner.js:257:21
    at _fulfilled (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\q\q.js:797:54)
    at self.promiseDispatch.done (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\q\q.js:826:30)
    at Promise.promise.promiseDispatch (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\q\q.js:759:13)
    at c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\q\q.js:573:44
    at flush (c:\Users\pcuser\AppData\Roaming\npm\node_modules\protractor\node_modules\q\q.js:108:17)
    at process._tickCallback (node.js:419:13)
[launcher] Process exited with error code 1
```

For only two files, everything works fine. For 3 files or more the error is thrown.

**Comment by juliemr** [comment]
I cannot reproduce this. Testing with 3 files. Can you reproduce reliably? Maybe this is a windows bug, has anyone else seen it?

**Comment by mm7163** [comment]
I can reproduce with two different test sets (first one is the one described above), on two different environments (although, both windows 7 sp1 x64). I uploaded my test set here https://github.com/mm7163/protractorRBBTIssue 
Please let me know if you need any additional details.

**Comment by cjroth** [comment]
+1

**Comment by juliemr** [comment]
Hmm Maybe a windows issue?

**Comment by tmulry** [comment]
I reproed this on MacOS Yosemite in chromedriver_2.15, selenium 2.45, protractor 2.1.0.  To avoid this error, you must remove all JS code from the describe block that is not part of a function declaration, and do not use any variables in your it() tests that you declared inside the "beforeEach" block.

I posted a refactored example at this fork https://github.com/tmulry/protractorRBBTIssue

FWIW, this error happens when I run tests locally or via SauceLabs.  

If you run `protractor config.js` from that repo, they will all pass.  Here is a sample refactor of test1.js (same goes for 2 and 3): 
    var m2generalObjects = require('../pageObjects/generalObjects.js');

    describe('Test Error', function() {
        
        beforeEach(function () {
          browser.driver.manage().window().maximize();
        });

        it('Should display correct content', function() {
            var ptor;
            ptor = browser;
            ptor.manage().deleteAllCookies();
            ptor.get(ptor.baseUrl);
            expect(element(m2generalObjects.text1).getText()).toContain('address');
        });
    });

**Comment by chris31389** [comment]
I had this issue too using cucumber-js.  I found that I needed to make sure that will every step function I included a callback and ended the step by calling the callback.  

`this.Then(/^I should see something$/, function(){ ... }`

becomes 

`this.Then(/^I should see something$/, function(done){ ...; done(); }`

**Comment by sjelin** [comment]
I have also reproduced this error on ubuntu

**Comment by prabmand** [comment]
I am facing this issue with OS x Yosemite too. After running the first test, a fresh browser session is not getting started. Tried the solution given by @tmulry but no result.

**Comment by mm7163** [comment]
The solution given by @tmulry worked for me. Thanks!

**Comment by rfodge** [comment]
Is there any update on this issue?  Occurring for me as well

**Comment by calendir** [comment]
I've had it happen to me while using Protractor 3.0.0 + CucumberJS 0.9.2 + Selenium Standalone 2.48.0 on Fedora 20. Would be nice to have an update on this...

**Comment by s-kavanagh** [comment]
I was also struggling with this for ages - @tmulry your workaround got things going for me! Thanks.

**Comment by szaluk** [comment]
Any update on this?  I can't get it working.

**Comment by Mras** [comment]
Same issue here. I am using page-objects and i was trying to adopt @tmulry `s solution, but it still fails locally and on browserstack:

```javascript
var helper = require('./helpers/webdriver_helper.js');
var myMainPage= require('./pageobjects/my_page.js');

var myTestFlow = function () {
myMainPage
    .navigate()
    .search(prodInfo.isbn);
}
describe('my site', function () {

  it('should go through flow in ".col-xs-$ - Extra Small" size', function (done) {
    helper.setDriverSize('S');
    myTestFlow ();
    done();
  });

  it('should go through flow in ".col-sm-$ - Small Devices" size', function (done) {
    helper.setDriverSize('M');
    myTestFlow ();
    done();
  });

  afterEach(function (done) {
    browser.driver.manage()
      .deleteAllCookies()
      .then(done);
  });
});

```
Am I missing something?

Feels like it fails to assign new value for browser.driver, so it should be easy to fix, right?

**Comment by leegee** [comment]
Same problem here.

Selenium v2.52.0, with Core v2.52.0. Built from revision 4c2593c
Driver info: driver.version: RemoteWebDriver
OS: Windows 10 10.0 amd64
Java: Oracle Corporation 25.91-b14
chromedriver_2.21 and chromedriver_2.15.322448 seem to act the same way
CEF 3.2454.1344.g2782fb8
Chromium 45.0.2454.101
WebKit 537.36
JavaScript 4.5.103.35

**Comment by MathieuMathurin** [comment]
I had the same error when using element.all in a page object file. 
The error was thrown when instantiating the page object once and using the tabs promise in more than one test. 
Fixed the issue by instantiating the tabs promise in the before each instead of in the page object file.

somePage.pageObjects.js
-------------------------------------------------------------------------------
var OtherObject = require(global.Root + '/Components/otherObject.js');

var pageObjects = function(){
    this.get = function(){
        browser.get(browser.baseUrl + "/en-CA/testPageURL");
    }

    //Line causing the issue
    this.OtherObject = new OtherObject();

    //Line fixing the issue
    this.OtherObject = OtherObject;
}

module.exports = new pageObjects();

-------------------------------------------------------------------------------

otherObject.js
-------------------------------------------------------------------------------
var OtherObject = function(){
    this.tabs = element(by.css(".mediazoneTabs")).all(by.repeater("tab in $context.tabs"));
}

module.exports = OtherObject;

-------------------------------------------------------------------------------

somePage.spec.js
-------------------------------------------------------------------------------
var pageObjects = require('../somePage.pageObjects.js');

//Line causing the issue
var OtherObject = pageObjects.OtherObject;

//Fix part 1: declare variable before
var OtherObject;

describe('When app is loaded', function () {

    beforeEach(function () {        
        //Fix part 2: instantiate object in beforeEach
        OtherObject = new pageObjects.OtherObject();

        //Get the page
        pageObjects.get();
    });

    it('should contains 3 tabs', function () {
        expect(OtherObject.tabs.count()).toBe(3);
    });

    it('Should have photo tab selected', function(){
        expect(OtherObject.tabs.first().getAttribute("class")).toContain("selected");
    });

});

-------------------------------------------------------------------------------

**Comment by holdfenytolvaj** [comment]
Same here if i want to have a restart only for a suit
i call 
    beforeAll(function () {
         browser.restart();
         browser.controlFlow().execute(...); //works fine;
    })

    it('', function () { 
          browser.controlFlow().execute(...); //throws error;
    })

**Comment by pedrin1001** [comment]
I am experiencing the same here.
* Node Version: 4.3.2
* Protractor Version: 3.2.2
* Browser(s): Chrome
* Operating System and Version openSUSE 13.2 (Harlequin) (x86_64) 64-bit

**Comment by juliemr** [comment]
Sorry for the delays here. On the radar!

**Comment by dogmatic69** [comment]
Same issue for me, on Ubuntu

```
/path/to/app/node_modules/selenium-webdriver/error.js:27
    super(opt_error);
    ^
UnsupportedOperationError: This driver instance does not have a valid session ID (did you call WebDriver.quit()?) and may no longer be used.
    at WebDriverError (/path/to/app/node_modules/selenium-webdriver/error.js:27:5)
    at UnsupportedOperationError (/path/to/app/node_modules/selenium-webdriver/error.js:436:5)
    at checkHasNotQuit (/path/to/app/node_modules/selenium-webdriver/lib/webdriver.js:378:15)
    at WebDriver.schedule (/path/to/app/node_modules/selenium-webdriver/lib/webdriver.js:323:5)
    at WebDriver.findElements (/path/to/app/node_modules/selenium-webdriver/lib/webdriver.js:897:22)
    at /path/to/app/node_modules/protractor/built/element.js:141:44
    at Promise.invokeCallback_ (/path/to/app/node_modules/selenium-webdriver/lib/promise.js:1329:14)
    at TaskQueue.execute_ (/path/to/app/node_modules/selenium-webdriver/lib/promise.js:2790:14)
    at TaskQueue.executeNext_ (/path/to/app/node_modules/selenium-webdriver/lib/promise.js:2773:21)
    at asyncRun (/path/to/app/node_modules/selenium-webdriver/lib/promise.js:2652:27)
    at /path/to/app/node_modules/selenium-webdriver/lib/promise.js:639:7
```

```
14:02 $ node --version
v6.3.0

14:06 $ npm --version
3.10.5

14:06 $ node_modules/.bin/protractor --version
Version 3.3.0
```

**Comment by jan-molak** [comment]
Same problem:

```
λ node --version
v4.4.5
λ npm --version
2.15.5
λ node_modules/.bin/protractor --version
Version 3.3.0
λ uname -a
Darwin jans-mbp.mynet 15.6.0 Darwin Kernel Version 15.6.0: Thu Jun 23 18:25:34 PDT 2016; root:xnu-3248.60.10~1/RELEASE_X86_64 x86_64
```

stacktrace:

```
[00:55:44] I/local - Shutting down selenium standalone server.

/path/to/app/node_modules/protractor/node_modules/selenium-webdriver/error.js:27
    super(opt_error);
         ^
UnsupportedOperationError: This driver instance does not have a valid session ID (did you call WebDriver.quit()?) and may no longer be used.
    at WebDriverError (/path/to/app/node_modules/protractor/node_modules/selenium-webdriver/error.js:27:10)
    at UnsupportedOperationError (/path/to/app/node_modules/protractor/node_modules/selenium-webdriver/error.js:436:10)
    at checkHasNotQuit (/path/to/app/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver.js:378:15)
    at /path/to/app/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver.js:349:7
    at TaskQueue.execute_ (/path/to/app/node_modules/protractor/node_modules/selenium-webdriver/lib/promise.js:2790:14)
    at TaskQueue.executeNext_ (/path/to/app/node_modules/protractor/node_modules/selenium-webdriver/lib/promise.js:2773:21)
    at /path/to/app/node_modules/protractor/node_modules/selenium-webdriver/lib/promise.js:2652:27
    at /path/to/app/node_modules/protractor/node_modules/selenium-webdriver/lib/promise.js:639:7
    at process._tickCallback (node.js:369:9)
From: Task: WebDriver.manage().timeouts().setScriptTimeout(110000)
    at WebDriver.schedule (/path/to/app/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver.js:344:17)
    at Timeouts.setScriptTimeout (/path/to/app/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver.js:1274:25)
    at Runner.createBrowser (/path/to/app/node_modules/protractor/built/runner.js:201:40)
    at Protractor.browser_.forkNewDriverInstance (/path/to/app/node_modules/protractor/built/runner.js:206:40)
    at Protractor.browser_.restart (/path/to/app/node_modules/protractor/built/runner.js:220:33)
    at Runner.restartDriver (/path/to/app/node_modules/protractor/built/runner.js:302:60)
    at emitOne (events.js:82:20)
    at Runner.emit (events.js:169:7)
    at /path/to/app/node_modules/protractor-cucumber-framework/lib/resultsCapturer.js:51:20
```

**Comment by jlin412** [comment]
+1

**Comment by marcincharezinski** [comment]
I dunno if it helps, but this one below works for me:
```
import {protractor, browser, element, by, By, $, $$, ExpectedConditions} from 'protractor/globals';

describe('Test works', () => {
    it('Go to angular', () => {
        protractor.browser.get("https://angularjs.org/");
    });
    it('Go to angular-tutorial', () => {
        protractor.browser.get("https://docs.angularjs.org/tutorial");
    });
});
```
If I change the code like:

```
import {browser, element, by, By, $, $$, ExpectedConditions} from 'protractor/globals';

describe('Test doesn't work', () => {
    it('Go to angular', () => {
       browser.get("https://angularjs.org/");
    });
    it('Go to angular-tutorial', () => {
        browser.get("https://docs.angularjs.org/tutorial");
    });
});
```
Protractor (Selenium) throws `UnsupportedOperationError: This driver instance does not have a valid session ID`

My conf.ts:

```
import {Config} from 'protractor';
export let config: Config = {
  seleniumAddress: 'http://localhost:4444/wd/hub',

  specs: ['spec_test.js'],
  restartBrowserBetweenTests: true
}
```

Using protractor 4.0.3, tested on both el capitan and ubuntu 14.04 TLS.

**Comment by Saurabh06** [comment]
+1 still facing same issue with protractor 3.x and on windows plateform

**Comment by jan-molak** [comment]
Again, same problem with Protractor 4.0.4. Is there anything we can do to help resolve this issue? Being able to restart the browser between tests is quite important... Thanks!

**Comment by juliemr** [comment]
OK, this issue has gotten too bloated and I think there are different issues in here, some of which have actually been fixed.

I think that the cleanups at https://github.com/angular/protractor/commit/3844eadaa1bba19d08133d613574c4fa730d1976 should help fix some of the remaining issues. That will be included in the next release.

For now, I'm going to close this one since it's not helpful for us anymore, and I think confusing bug searchers more than helping them. But, if you have a reproducible test case with the latest protractor/selenium, I would love to see it in a new issue. Thanks!

**Comment by dhinakarankannan** [comment]
Even i have the same issue. When running multiple tests in mocha command line getting this error: "![CDATA[Error: timeout of 99000000ms exceeded. Ensure the done() callback is being called in this test". but when done individually or when done into parts instead of running all together not getting any error.

**Comment by fescobar** [comment]
I have the solution, just remove Protractor from your project. I don't understand what is the benefit to use this "framework" when you can use directly **selenium webdriver**. I have to deal with multiple issues only because one of my current projects implement this thing. I think the committers need to learn more about software architecture before starting to build any tool. Good luck!

**Comment by simonua** [comment]
@fescobar, respectfully, I'm not sure being snarky is productive.

**Comment by prashanth-sams** [comment]
Indeed very important feature and this task needs to be re-opened

**Comment by jan-molak** [comment]
Hi all, I've solved the issue with browser restarts in [Serenity/JS](https://serenity-js.org/) (which [integrates with Protractor](https://serenity-js.org/modules/protractor)).

Try the following if you're still seeing it in your Protractor tests:

Add Serenity/JS core and the Protractor adapter to your project (you can also add a console reporter if you wish):

```
npm install --save-dev @serenity-js/core @serenity-js/protractor @serenity-js/console-reporter
```
Add one of the below test runner adapters, so if you're using Jasmine:
```
npm install --save-dev @serenity-js/jasmine
```
If you're using Mocha:
```
npm install --save-dev @serenity-js/mocha
```
If you're using Cucumber:
```
npm install --save-dev @serenity-js/cucumber
```

Next, modify `protractor.conf.js` to use Serenity/JS between Protractactor and your test runner (see [more detailed example](https://serenity-js.org/modules/protractor/)):

```javascript
const { ConsoleReporter } = require('@serenity-js/console-reporter');

exports.config = {
    // Tell Protractor to use the Serenity/JS framework Protractor Adapter
    framework:      'custom',
    frameworkPath:  require.resolve('@serenity-js/protractor/adapter'),

    // Configure Serenity/JS to use an appropriate test runner
    // and the Stage Crew Members we've imported at the top of this file
    serenity: {
        runner: 'jasmine',
        // runner: 'cucumber',
        // runner: 'mocha',

        // optional, only if you want to use Serenity/JS reporters
        crew: [
            ConsoleReporter.withDefaultColourSupport()
        ]
    },

    // configure Cucumber runner
    cucumberOpts: {
        // see the Cucumber configuration options below
    },

    // or configure Jasmine runner
    jasmineNodeOpts: {
        // see the Jasmine configuration options below
    },

    // or configure Mocha runner
    mochaOpts: {
        // see the Mocha configuration options below
    },

    // ... other Protractor-specific configuration   
};
```

And here are integrations tests that verify restarting the browser works:
- with [Jasmine](https://github.com/serenity-js/serenity-js/blob/master/integration/protractor-jasmine/spec/restarting_browser_between_tests.spec.ts)
- with [Mocha](https://github.com/serenity-js/serenity-js/blob/master/integration/protractor-mocha/spec/restarting_browser_between_tests.spec.ts)

Also, please note my comment in this thread from [4 years ago](https://github.com/angular/protractor/issues/2001#issuecomment-243912191). Using `protractor.browser` instead of `browser` in your tests _does_ make a difference.

Hope this helps!
Jan
